### ISSUE TITLE
Major flaw in repeater's design: pagination div is part of the items list!

### ISSUE THREAD

**Issue by yankiara** [soflyy/oxygen-bugs-and-features#959]
**Describe the bug**
The pagination div in a repeater is the last item of the repeater's list. This is wrong, the pagination SHOULD NOT be part of items list.
This should be obvious by design, but the real issue is that it messes with CSS rules when applying to repeater's list or items.
For instance, using flexbox properties or grid affects the pagination div and destroys layout.

**Have you tried all the steps at https://oxygenbuilder.com/documentation/troubleshooting/troubleshooting-guide/?**
Yes

**Are you able to replicate the issue on a Sandbox install at https://oxygenbuilder.com/try?**
https://oxygen-qn52kdaf76v8k.oxygen-demo.qsandbox.me/sample-page/

**To Reproduce**
1. Go to any page
2. Add a repeater with a any post query (at least one post needed to check the result)
3. Apply 2 or 3 columns grid to repeater or flexbox 50% width and flex-wrap
4. See that pagination is in a column in the last row instead of at bottom of the list

**Expected behavior**
Pagination div should be at the bottom of the list, AFTER the list, out ouf the .oxy-dynamic-list div.

**What actually happens**
Pagination div is inside the list, so it is part of the columns!

**Screenshots**
![repeater pagination](https://user-images.githubusercontent.com/36680757/78471524-36d1a300-7732-11ea-8ece-f31666e3c0ad.png)

**Desktop (please complete the following information):**
 - Oxygen Version 3.2.1

**Comment by KittenCodes** [comment]
You can move the pagination below the repeater with CSS:

```css
.oxy-repeater-pages-wrap {
	width: 100%;
}
```

**Comment by yankiara** [comment]
OK, but semantically speaking, pagination should not be a element of the repeater's list.
What about dynamically ordering the list or any other process in JS?
The pagination div parasites all CSS or jQuery find.

**Comment by d-packs** [comment]
> You can move the pagination below the repeater with CSS:
> 
> ```css
> .oxy-repeater-pages-wrap {
> 	width: 100%;
> }
> ```
That's not really a solution and this problem is real. Pagination should definitely not be treated like any other item in the list. It's a concept flaw really.

**Comment by rickfitz** [comment]
This completely messes up putting the repeater in a carousel like Flickity. There's always one extra slide that shouldn't be there. So looks like I now have to add extra js to delete the last element every time :(

**Comment by roireshef** [comment]
Guys (@KittenCodes) - any clean solution?

**Comment by KittenCodes** [comment]
> Guys (@KittenCodes) - any clean solution?

Only the above CSS.

**Comment by yankiara** [comment]
Sometimes, there's no other choice than removing it from the DOM, for instance:
```
jQuery('.your-slider .oxy-repeater-pages-wrap').remove();
```

**Comment by warm--tape** [comment]
Definitely a major issue. Tried applying CSS grid to a repeater. Should be simple enough, but no- layout ruined because pagination becomes one of the grid items and adds extra spacing at the bottom of the grid.
<img width="1148" alt="Screen Shot 2020-07-17 at 4 44 23 PM" src="https://user-images.githubusercontent.com/10751100/87749442-cd8f1f00-c84c-11ea-9b70-8133c8f2d279.png">
<img width="558" alt="Screen Shot 2020-07-17 at 4 44 08 PM" src="https://user-images.githubusercontent.com/10751100/87749449-d08a0f80-c84c-11ea-8666-4ff84ade3630.png">

**Comment by KittyFolks** [comment]
Is this going to be addressed soon? I have tons of repeaters all across my site and there is no chance I switch them all to manual to add "no_found_rows" manually and rewrite the whole query. The whole point of using a repeater is lost. The issue is that I have defined a max number of posts, which should remove the pagination automatically. As I understand, the pagination itself is added as another element and thats why the pagination shows up. 
I have hidden the element through CSS but this doesn't remove it from the DOM of course.
There should be an option to disable pagination in custom mode. 

Is there a function I can use to accomplish the removal of .oxy-repeater-pages-wrap until this is fixed?

**Comment by swinggraphics** [comment]
> Is this going to be addressed soon?

No. See the conversations about duplicate IDs and other bad coding practices. Once they release something, they stick with it.

**Comment by YammYcoding** [comment]
this would be so easy to fix, just by adding a counter variable after the id name.. like...

`for ( $i=1; $i < $list_size; $i++ ) {
     print "<article id="$divname". "_" . $i;
}`

**Comment by roireshef** [comment]
I honestly didn't expect them to ignore this simple request and to not provide any clean solution for so much time, so I'm going to share the 10sec solutions I came across digging into their code, specifically into `wp-content/plugins/oxygen/component-framework/components/classes/dynamic-list.class.php` just comment out this whole code block and this damn repeater pages element will be gone from your life:

```
                     ?>
                     <div class="oxy-repeater-pages-wrap">
                         <div class="oxy-repeater-pages">
                         <?php
                         $pagination_args = array(
                             'base' => str_replace( $big, '%#%', esc_url( get_pagenum_link( $big ) ) ),
                             'format' => '?paged=%#%',
                             'current' => max( 1, get_query_var('paged') ),
                             'total' => $query->max_num_pages,

                         );

                         if(isset($options['paginate_mid_size']) && is_numeric($options['paginate_mid_size'])) {
                             $pagination_args['mid_size'] =  intval($options['paginate_mid_size']);
                         }

                         if(isset($options['paginate_end_size']) && is_numeric($options['paginate_end_size'])) {
                             $pagination_args['end_size'] =  intval($options['paginate_end_size']);
                         }

                         if(isset($options['paginate_prev_link_text']) && $options['paginate_prev_link_text']) {
                             $pagination_args['prev_text'] =  sanitize_text_field(oxygen_vsb_base64_decode($options['paginate_prev_link_text']));
                         }
                         if(isset($options['paginate_next_link_text']) && $options['paginate_next_link_text']) {
                             $pagination_args['next_text'] =  sanitize_text_field(oxygen_vsb_base64_decode($options['paginate_next_link_text']));
                         }

                         echo paginate_links( $pagination_args );
                        ?>
                         </div>
                     </div>
                    <?php

```

I must say, not coming from the web-development domain, I can't even call reasonable covering it up with CSS to hide a redundant element, or use JS code to remove it. It shouldn't be created in the first place.

**Comment by wpsumo** [comment]
> this would be so easy to fix, just by adding a counter variable after the id name.. like...
> 
> `for ( $i=1; $i < $list_size; $i++ ) { print "<article id="$divname_$i"; ( ... ) }`
> 
> I don't know why this was not added to priority yet....

@YammYcoding I agree, or just move everything to classes.  And solve the issue where we are not forced to put styling int he global universal.css stylesheet as soon as we use classes.

**Comment by jamescrabb** [comment]
And whilst they are at it they can fix the duplicated id issue with repeaters!

**Comment by Jrodriguez1001** [comment]
@jamescrabb you can use this code for to fix the duplicated id
 `(function($) {

    //abort if we're viewing inside builder
    if ($('html').attr('ng-app') == 'CTFrontendBuilder')
        return;

    var ids = {};
    $('.oxygen-body [id]').each(function() {
        if (this.id && ids[this.id]) {
            $(this).removeAttr('id');
        }
        ids[this.id] = 1;
    });

    $('.rep-carousel').flickity({
        cellSelector: '.rep-carousel-cell',
    });

})(jQuery);`

**Comment by louisreingold** [comment]
Not sure if this solution would work, but...

What if we just have an option to not output IDs if an element is inside a Repeater? The option could be a checkbox on the Repeater, and not enabled by default. 

Enabling it obviously would mean your ID styles for elements inside a Repeater wouldn't work. But that's not a problem, since people who want to enable this option are going to use classes to style the Repeater's children.

Over the last year we've done a fantastic job fixing bugs and making Oxygen significantly more reliable. So let's compromise a little bit on code quality and ship this. It's not like this is the only bit of partially-digested spaghetti in the codebase.

When rendering a Repeater...
```php
global $isInRepeater;
$isInRepeater = true;
/* render contents */
$isInRepeater = false;
```

When outputting any element's HTML on the frontend...
```php
global $isInRepeater

/* when outputting the HTML ID attributes, wrap with... */
if (!$isInRepeater) {
 // ...
}
```

Is there something about the codebase that prevents a "hacky" solution like this or makes implementation more complicated than what I've described above?

@ilyakop @Spellhammer

I won't be commenting further.

**Comment by louisreingold** [comment]
Someone's gonna find this post in 5 years and accuse me of being a shitty developer. 🤷‍♀️

**Comment by Jrodriguez1001** [comment]
You are right. I will delete my comment.

**Comment by stefanullinger** [comment]
Hi guys, so I was just running in the same issue. Actually I switched from Elementor to Oxygen because 1) Elementor has many issues and 2) many people recommended to try Oxygen. But now that I see that Oxygen also has so many issues, I will definitely search for a better solution.

For now, my quick fix for this issue with my 3-column grid is:
```
.oxy-repeater-pages-wrap {
  grid-column: span 3;
}
```

**Comment by yankiara** [comment]
And for a responsive grid (either with breakpoint or auto-fit), this is working:
.oxy-repeater-pages-wrap {
  grid-column: 1 / -1;
}

**Comment by JoeKns** [comment]
This is still an issue in May 2021. I honestly don't understand why.

**Comment by d-packs** [comment]
I think it's because once released, changing the HTML output of a component would break sites that are using that component. If it was horrible code upon release, than that's what we're stuck with. So your best option is to either catch the shortcode output via do_shortcode filter, or javascript to move the pagination element out of the container.

**Comment by swinggraphics** [comment]
It'd be very easy enough keep the existing code for existing elements and use corrected code for new instances. They just don't care much about standards and best practices. If they did, stuff like this would never make it out the door to begin with. I've been admonished for making declarations like that before, but it is evidenced in every corner of their product.

**Comment by v-marinkov** [comment]
This must be fixed. Yet another simple issue completely ignored

**Comment by KonradDeskiewicz** [comment]
I believe there won't be any changes until Oxy 4.0 (then braking changes are expected).

**Comment by paulmatunog** [comment]
Came here because of this issue. The code below works for me.

```
.oxy-repeater-pages-wrap {
  grid-column: 1 / -1;
}

```

**Comment by bocohostwill** [comment]
> Came here because of this issue. The code below works for me.
> 
> ```
> .oxy-repeater-pages-wrap {
>   grid-column: 1 / -1;
> }
> ```

This worked for me as well

**Comment by JoeKns** [comment]
It's simply a joke this is actually still a thing after all this time.

**Comment by JoeKns** [comment]
> I believe there won't be any changes until Oxy 4.0 (then braking changes are expected).

Wrong. As sad as that is, it's not a major update. We got a few minor fixes and a few minor UI improvements.

**Comment by reachingmike** [comment]
+1

**Comment by drakedeatonuk** [comment]
+1

**Comment by Stoxis** [comment]
Why hasn't this been added...

**Comment by 4tcmedia** [comment]
This actually still not added - 2024 now 
This worked for me with auto fit columns in grid (4 rows)
.oxy-repeater-pages-wrap {
  grid-column: 1 / -1;
}

**Comment by ouengai** [comment]
This is still an issue in May 2026. I honestly don't understand why.
